mirrord/layer/src/file/hooks.rs

```diff
@@ -181,6 +181,154 @@ pub(super) unsafe extern "C" fn opendir_detour(raw_filename: *const c_char) -> u
+            libc::read(in_fd, buffer.as_mut_ptr() as *mut c_void, count)
+        }
+    };
+
+    if bytes_read < 0 {
+        return None;
+    }
+
+    let written = unsafe {
+        libc::write(
+            out_fd,
+            buffer.as_ptr() as *const c_void,
+            bytes_read as usize,
+        )
+    };
+
+    if written < 0 {
+        return None;
+    }
+
+    Some((written, offset.unwrap_or(0) + bytes_read as off_t))
+}
+
+/// Hook for macos's [`libc::sendfile`].
+#[cfg(target_os = "macos")]
+#[hook_fn]
+pub(super) unsafe extern "C" fn sendfile_macos_detour(
+    fd: c_int,
+    s: c_int,
+    offset: off_t,
+    len: *mut off_t,
+    _hdtr: *const libc::sf_hdtr,
+    _flags: c_int,
+) -> c_int {
+    unsafe {
+        let Some(count) = len.as_mut() else {
+            return -1;
+        };
+
+        sendfile_impl(fd, s, Some(offset), *count as usize)
+            .map(|(written, _)| {
+                *count = written as off_t;
+                0
+            })
+            .unwrap_or(-1)
+    }
+}
+
+/// Hook for linux's [`libc::sendfile`].
+#[cfg(target_os = "linux")]
+#[hook_fn]
+pub(super) unsafe extern "C" fn sendfile_linux_detour(
+    out_fd: c_int,
+    in_fd: c_int,
+    offset: *mut off_t,
+    count: size_t,
+) -> ssize_t {
+    unsafe {
+        let offset_val = if offset.is_null() {
+            None
+        } else {
+            Some(*offset)
+        };
+
+        sendfile_impl(in_fd, out_fd, offset_val, count)
+            .map(|(written, new_offset)| {
+                if !offset.is_null() {
+                    *offset = new_offset;
+                }
+                written
+            })
+            .unwrap_or(-1)
+    }
+}
+
+/// Hook for [`libc::ftruncate`].
+#[hook_guard_fn]
+pub(super) unsafe extern "C" fn ftruncate_detour(fd: c_int, length: off_t) -> c_int {
+    ftruncate(fd, length)
+        .map(|()| 0)
+        .unwrap_or_bypass_with(|_| unsafe { FN_FTRUNCATE(fd, length) })
+}
+
+/// Hook for [`libc::futimens`].
+#[hook_guard_fn]
+pub(super) unsafe extern "C" fn futimens_detour(fd: c_int, raw_times: *const timespec) -> c_int {
+    unsafe {
+        let times = if !raw_times.is_null() {
+            let [first, second] = slice::from_raw_parts(raw_times, 2) else {
+                unreachable!("We create the slice with two elements")
+            };
+
+            Some([
+                Timespec {
+                    tv_sec: first.tv_sec,
+                    tv_nsec: first.tv_nsec,
+                },
+                Timespec {
+                    tv_sec: second.tv_sec,
+                    tv_nsec: second.tv_nsec,
+                },
+            ])
+        } else {
+            None
+        };
+        futimens(fd, times)
+            .map(|()| 0)
+            .unwrap_or_bypass_with(|_| FN_FUTIMENS(fd, raw_times))
+    }
+}
+
+/// Hook for [`libc::fchown`].
+#[hook_guard_fn]
+pub(super) unsafe extern "C" fn fchown_detour(fd: c_int, owner: uid_t, group: gid_t) -> c_int {
```

Comment:
how is the `fchown` function used by node's `copyFile()`? I assume it copies the src file's owner and set them for the dst file?